riscv/src/continuations/memory_merkle_tree.rs

```diff
@@ -0,0 +1,272 @@
+use std::collections::BTreeMap;
+
+use crate::bootloader::{
+    BYTES_PER_WORD, MERKLE_TREE_DEPTH as MERKLE_TREE_DEPTH_BOOTLOADER,
+    WORDS_PER_PAGE as WORDS_PER_PAGE_BOOTLOADER,
+};
+use number::FieldElement;
+use riscv_executor::poseidon_gl::poseidon_gl;
+
+/// A Merkle tree of memory pages.
+pub struct MerkleTree<
+    T: FieldElement,
+    const MERKLE_TREE_DEPTH: usize = MERKLE_TREE_DEPTH_BOOTLOADER,
+    const WORDS_PER_PAGE: usize = WORDS_PER_PAGE_BOOTLOADER,
+> {
+    /// Hashes of all nodes of the Merkle tree. The vector has MERKLE_TREE_DEPTH + 1 entries
+    /// where hashes[0] is [root_hash], hashes[1] is [level_1_hash_0, level_1_hash_1], etc.
+    /// The last entry hashes[MERKLE_TREE_DEPTH] is the hash of the leaves.
+    /// These hashes should be updated whenever the data is updated.
+    hashes: Vec<Vec<[T; 4]>>,
+    /// Memory pages.
+    data: Vec<[T; WORDS_PER_PAGE]>,
+}
+
+/// Computes the Poseidon hash of two 4-field-element inputs, using a capacity of 0.
+fn hash_cap0<T: FieldElement>(data1: &[T], data2: &[T]) -> [T; 4] {
```

Comment:
Why is the size of `data1` and `data2` not enforced here? I'm wondering because the code below uses at most 4 elements from each